neuro_san/internals/graph/activations/abstract_class_activation.py

```diff
@@ -234,24 +234,11 @@ async def attempt_invoke(self, coded_tool: CodedTool, arguments: Dict[str, Any],
         try:
             # Try the preferred async_invoke()
             retval = await coded_tool.async_invoke(self.arguments, self.sly_data)
-        except NotImplementedError:
-            # That didn't work, so try running the synchronous method as an async task
-            # within the confines of the proper executor.
 
-            # Warn that there is a better alternative.
-            message = f"""
-Running CodedTool class {coded_tool.__class__.__name__}.invoke() synchronously in an asynchronous environment.
-This can lead to performance problems when running within a server. Consider porting to the async_invoke() method.
-"""
-            self.logger.info(message)
-            message = AgentMessage(content=message)
-            await self.journal.write_message(message)
-
-            # Try to run in the executor.
-            invocation_context = self.run_context.get_invocation_context()
-            executor: AsyncioExecutor = invocation_context.get_asyncio_executor()
-            loop: AbstractEventLoop = executor.get_event_loop()
-            retval = await loop.run_in_executor(None, coded_tool.invoke, arguments, sly_data)
+        except NotImplementedError:
+            # That didn't work, so try running the synchronous method
+            # in a separate thread so as not to distrupt the main event loop.
+            retval = await to_thread(coded_tool.invoke, arguments, sly_data)
```

Comment:
Instead of complaining about the user's code, send any specifically synchronous invoke() execution to a separate thread so the main event loop thread will not be blocked.

Reply:
Nice higher-level wrapper API method, which should be functionally the same. If we are in the running env of our request AsyncioExecutor. But that should be the case here, so we are sending the workload to the same event loop.

Reply:
That's a very good thought in general, but: if we are talking about potential concurrency issues, we need at least 2 threads to have some concurrency issues. Right? So one thread is started by asyncio.to_thread() OK. What is another one? If it is our per-request thread we run in our AsyncioExecutor, than it actually only runs our (per-request) event loop.

Reply:
@andreidenissov-cog : The next thread would be called by the next synchronous call to any CodedTool.invoke().

@ofrancon 's point is a very subtle but good one. I am going to revert this new logic in favor of documentation for the to_thread() scenario.